src/ui/profile/ProfileSelection.js

```diff
@@ -0,0 +1,170 @@
+// eslint-disable-next-line no-redeclare
+/* global chrome */
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import styles from '../styles/profile.module.css';
+import AddProfile from './AddProfile';
+import BeyondLoader from './BeyondLoader';
+import Profiles from './Profile';
+
+const LoadedProfile = (profileModel) => ({
+  status: 'loaded',
+  profile: profileModel,
+});
+
+const ErrorProfile = (error, profile) => ({
+  status: 'error',
+  error,
+  profile,
+});
+
+const LoadingProfile = (profile) => ({
+  status: 'loading',
+  profile,
+});
+
+function ProfileSelection(props) {
+  const { onCharacterReady } = props;
+  const [profiles, setProfiles] = useState(new Map());
+  const [currentProfile, setCurrentProfile] = useState(null);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    chrome.storage.local.get('savedProfiles', (result) => {
+      if (!result.savedProfiles) return;
+
+      const keyPairs = result.savedProfiles.map((profile) => [profile.id, LoadedProfile(profile)]);
+      setProfiles(new Map(keyPairs));
+    });
+  }, []);
+
+  const onAddProfile = (id, profilePromise) => {
+    if (profiles.has(id)) {
+      throw new Error('ALREADY_EXISTS');
+    }
+
+    // Add placeholder profile with id until loaded
+    setProfiles((map) => {
+      const newMap = new Map(Array.from(map.entries()));
+      newMap.set(id, LoadingProfile({ id }));
+      return newMap;
+    });
+
+    profilePromise.then((profile) => {
+      setProfiles((map) => {
+        const newMap = new Map(Array.from(map.entries()));
+        newMap.set(id, LoadedProfile(profile));
+
+        // Append the loaded profile to list of saved profiles
+        chrome.storage.local.get('savedProfiles', (result) => {
+          const savedList = result.savedProfiles ? [...result.savedProfiles, profile] : [profile];
```

Comment:
Array spread syntax will gracefully handle `null` and `undefined`, so this can be simplified to:

```js
const savedList = [...result.savedProfiles, profile];
```